staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -773,24 +773,26 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
 		return err
 	}
 	span.AddEvent("Listed items from cache", attribute.Int("count", len(objs)))
-	if len(objs) > listVal.Cap() && pred.Label.Empty() && pred.Field.Empty() {
-		// Resize the slice appropriately, since we already know that none
-		// of the elements will be filtered out.
-		listVal.Set(reflect.MakeSlice(reflect.SliceOf(c.objectType.Elem()), 0, len(objs)))
```

Comment:
would the benchmarks be even better if you just omitted this and continued to use append?

Reply:
As suggested, I used the following code for testing: [https://github.com/sxllwx/kubernetes/blob/optimize/watch-cache-getlist-fix-comment/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go](https://github.com/sxllwx/kubernetes/blob/optimize/watch-cache-getlist-fix-comment/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go)

The test results are as follows:

```go
# cat cache.before                                                                                                              
goos: linux
goarch: amd64
pkg: k8s.io/apiserver/pkg/storage/cacher
cpu: AMD EPYC 7K62 48-Core Processor
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16         	     847	   1374085 ns/op	  156705 B/op	     132 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16         	     865	   1386726 ns/op	  156078 B/op	     130 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16        	     757	   1513711 ns/op	  626450 B/op	     588 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16        	     750	   1511499 ns/op	  626509 B/op	     588 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16       	     650	   1761892 ns/op	 1696343 B/op	    1095 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16       	     643	   1759985 ns/op	 1696426 B/op	    1096 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16       	     424	   2731813 ns/op	 5431142 B/op	    2619 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16       	     421	   2737815 ns/op	 5431237 B/op	    2619 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16       	     290	   4055868 ns/op	10330039 B/op	    5148 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16       	     286	   4007450 ns/op	10330295 B/op	    5149 allocs/op
PASS
ok  	k8s.io/apiserver/pkg/storage/cacher	34.838s

# cat cache.after
goos: linux
goarch: amd64
pkg: k8s.io/apiserver/pkg/storage/cacher
cpu: AMD EPYC 7K62 48-Core Processor
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16         	     850	   1360398 ns/op	  119860 B/op	      86 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16         	     872	   1364373 ns/op	  119206 B/op	      84 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16        	     817	   1430057 ns/op	  360647 B/op	      89 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16        	     806	   1431489 ns/op	  360690 B/op	      90 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16       	     748	   1491980 ns/op	  640999 B/op	      94 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16       	     747	   1484896 ns/op	  641051 B/op	      94 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16       	     663	   1703111 ns/op	 1456964 B/op	     100 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16       	     673	   1686134 ns/op	 1456867 B/op	     100 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16       	     568	   2003787 ns/op	 2834520 B/op	     109 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16       	     564	   2010633 ns/op	 2834591 B/op	     109 allocs/op
PASS
ok  	k8s.io/apiserver/pkg/storage/cacher	34.016s

# cat cache.comment                                                                                                            
goos: linux
goarch: amd64
pkg: k8s.io/apiserver/pkg/storage/cacher
cpu: AMD EPYC 7K62 48-Core Processor
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16         	     841	   1373561 ns/op	  156746 B/op	     132 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16         	     862	   1366341 ns/op	  156134 B/op	     130 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16        	     769	   1519386 ns/op	  626438 B/op	     587 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16        	     735	   1508256 ns/op	  626657 B/op	     589 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16       	     638	   1764863 ns/op	 1696476 B/op	    1096 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16       	     644	   1759715 ns/op	 1696403 B/op	    1096 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16       	     418	   2721730 ns/op	 5431312 B/op	    2620 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16       	     418	   2732375 ns/op	 5431319 B/op	    2620 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16       	     292	   4007642 ns/op	10329922 B/op	    5148 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16       	     291	   4042856 ns/op	10329993 B/op	    5148 allocs/op
PASS
ok  	k8s.io/apiserver/pkg/storage/cacher	34.775s

# benchcmp cache.before cache.after                                                                                                 
benchcmp is deprecated in favor of benchstat: https://pkg.go.dev/golang.org/x/perf/cmd/benchstat
benchmark                                                                old ns/op     new ns/op     delta
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       1374085       1360398       -1.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       1386726       1364373       -1.61%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      1513711       1430057       -5.53%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      1511499       1431489       -5.29%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1761892       1491980       -15.32%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1759985       1484896       -15.63%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     2731813       1703111       -37.66%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     2737815       1686134       -38.41%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     4055868       2003787       -50.60%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     4007450       2010633       -49.83%

benchmark                                                                old allocs     new allocs     delta
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       132            86             -34.85%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       130            84             -35.38%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      588            89             -84.86%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      588            90             -84.69%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1095           94             -91.42%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1096           94             -91.42%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     2619           100            -96.18%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     2619           100            -96.18%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     5148           109            -97.88%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     5149           109            -97.88%

benchmark                                                                old bytes     new bytes     delta
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       156705        119860        -23.51%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       156078        119206        -23.62%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      626450        360647        -42.43%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      626509        360690        -42.43%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1696343       640999        -62.21%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1696426       641051        -62.21%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     5431142       1456964       -73.17%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     5431237       1456867       -73.18%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     10330039      2834520       -72.56%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     10330295      2834591       -72.56%

# benchcmp cache.before cache.comment                                                                                                
benchcmp is deprecated in favor of benchstat: https://pkg.go.dev/golang.org/x/perf/cmd/benchstat
benchmark                                                                old ns/op     new ns/op     delta
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       1374085       1373561       -0.04%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       1386726       1366341       -1.47%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      1513711       1519386       +0.37%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      1511499       1508256       -0.21%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1761892       1764863       +0.17%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1759985       1759715       -0.02%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     2731813       2721730       -0.37%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     2737815       2732375       -0.20%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     4055868       4007642       -1.19%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     4007450       4042856       +0.88%

benchmark                                                                old allocs     new allocs     delta
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       132            132            +0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       130            130            +0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      588            587            -0.17%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      588            589            +0.17%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1095           1096           +0.09%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1096           1096           +0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     2619           2620           +0.04%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     2619           2620           +0.04%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     5148           5148           +0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     5149           5148           -0.02%

benchmark                                                                old bytes     new bytes     delta
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       156705        156746        +0.03%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       156078        156134        +0.04%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      626450        626438        -0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      626509        626657        +0.02%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1696343       1696476       +0.01%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1696426       1696403       -0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     5431142       5431312       +0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     5431237       5431319       +0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     10330039      10329922      -0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     10330295      10329993      -0.00%
```

Conclusion: no noticeable performance gain. (I guess: Golang does Copy internally during the Append process)

Reply:
Thank you for checking!